src/main/java/slimeknights/tconstruct/smeltery/recipe/AbstractCastingRecipe.java

```diff
@@ -0,0 +1,202 @@
+  protected final int coolingTime;
+  private final boolean consumed;
+  private final boolean switchSlots;
+
+  public AbstractCastingRecipe(IRecipeType<?> typeIn, ResourceLocation idIn, String groupIn,
+                               @Nullable Ingredient ingredient, @Nonnull FluidStack fluidIn, ItemStack result,
+                               int coolingTime, boolean consumed, boolean switchSlots) {
+    this.type = typeIn;
+    this.id = idIn;
+    this.group = groupIn;
+    this.cast = ingredient;
+    this.fluid = fluidIn;
+    this.result = result;
+    this.coolingTime = coolingTime;
+    this.consumed = consumed;
+    this.switchSlots = switchSlots;
+  }
+
+  @Override
+  public boolean matches(IInventory inv, World worldIn) {
+    return this.cast.test(inv.getStackInSlot(0));
+  }
+
+  public boolean matches(Fluid fluid, IInventory inv, World world) {
+    return (this.fluid.getFluid() == fluid || fluid == Fluids.EMPTY) && this.matches(inv, world);
+  }
+
+  @Override
+  public ItemStack getCraftingResult(IInventory inv) {
+    return result;
+  }
+
+  @Override
+  public boolean canFit(int width, int height) {
+    return true;
+  }
+
+  @Override
+  public ItemStack getRecipeOutput() {
+    return this.result;
+  }
+
+  @Override
+  public ResourceLocation getId() {
+    return this.id;
+  }
+
+  @Override
+  public IRecipeSerializer<?> getSerializer() {
+    return null;
+  }
+
+  @Override
+  public IRecipeType<?> getType() {
+    return type;
+  }
+
+  public int getCoolingTime() {
+    return this.coolingTime;
+  }
+
+  public int getFluidAmount() {
+    return this.fluid.getAmount();
+  }
+
+  public FluidStack getFluidStack() {
+    return this.fluid;
+  }
+
+  public Fluid getFluid() {
+    return this.fluid.getFluid();
+  }
+
+  public Ingredient getCast() {
+    return cast;
+  }
+
+  /**
+   * Does this recipe consume/destroy the cast ingredient.
+   * @return  Whether to destroy cast input.
+   */
+  public boolean consumesCast() {
+    return this.consumed;
+  }
+
+  /**
+   * Does this recipe's output get put into the input slot
+   * @return  Whether to switch to input after casting
+   */
+  public boolean switchSlots() {
+    return this.switchSlots;
+  }
+
+  /**
+   * Keeps casting recipes out of the recipe book.
+   */
+  @Override
+  public boolean isDynamic() {
+    return true;
+  }
+
+  public static class Serializer<T extends AbstractCastingRecipe> extends ForgeRegistryEntry<IRecipeSerializer<?>>
+    implements IRecipeSerializer<AbstractCastingRecipe> {
+    private final Serializer.IFactory<T> factory;
+
+    public Serializer(Serializer.IFactory<T> factoryIn) {
+      this.factory = factoryIn;
+    }
+
+    @Override
+    public T read(ResourceLocation recipeId, JsonObject json) {
+      Ingredient cast = Ingredient.EMPTY;
+      String s = JSONUtils.getString(json, "group", "");
+      boolean consumed = false;
+      boolean switchSlots = JSONUtils.getBoolean(json, "switchslots", false);
```

Comment:
I suggest using `switch_slots`, more readible.

Reply:
the variable or the json field?

Reply:
JSON field. variable is fine in camel case.

Reply:
Looking over this again, noticed a couple other fields in JSON like that. In general, JSON fields should be `snake_case` instead of `lowercase`